lib/Service/RecordingService.php

```diff
@@ -0,0 +1,140 @@
+ *
+ */
+
+namespace OCA\Talk\Service;
+
+use InvalidArgumentException;
+use OC\User\NoUserException;
+use OCA\Talk\Config;
+use OCA\Talk\Exceptions\ParticipantNotFoundException;
+use OCA\Talk\Room;
+use OCP\Files\Folder;
+use OCP\Files\IMimeTypeDetector;
+use OCP\Files\IRootFolder;
+use OCP\Files\NotFoundException;
+use OCP\Files\NotPermittedException;
+
+class RecordingService {
+	public const DEFAULT_ALLOWED_RECORDING_FORMATS = [
+		'audio/ogg' => ['ogg'],
+		'video/ogg' => ['ogv'],
+		'video/x-matroska' => ['mkv'],
+	];
+	private IMimeTypeDetector $mimeTypeDetector;
+	private ParticipantService $participantService;
+	private IRootFolder $rootFolder;
+	private Config $config;
+
+	public function __construct(
+		IMimeTypeDetector $mimeTypeDetector,
+		ParticipantService $participantService,
+		IRootFolder $rootFolder,
+		Config $config
+	) {
+		$this->mimeTypeDetector = $mimeTypeDetector;
+		$this->participantService = $participantService;
+		$this->rootFolder = $rootFolder;
+		$this->config = $config;
+	}
+
+	public function store(Room $room, string $owner, array $file): void {
+		$content = $this->getContentFromFileArray($file);
+
+		$this->validateFileName($file['name']);
+		$this->validateFileFormat($file['name'], $content);
+
+		try {
+			$this->participantService->getParticipant($room, $owner);
+		} catch (ParticipantNotFoundException $e) {
+			throw new InvalidArgumentException('owner_participant');
+		}
+
+		try {
+			$recordingFolder = $this->getRecordingFolder($owner, $room->getToken());
+			$recordingFolder->newFile($file['name'], $content);
+		} catch (NoUserException $e) {
+			throw new InvalidArgumentException('owner_invalid');
+		} catch (NotPermittedException $e) {
+			throw new InvalidArgumentException('owner_permission');
+		}
+	}
+
+	public function getContentFromFileArray(array $file): string {
+		if (
+			$file['error'] !== 0 ||
+			!is_uploaded_file($file['tmp_name'])
+		) {
+			throw new InvalidArgumentException('invalid_file');
+		}
+
+		$content = file_get_contents($file['tmp_name']);
+		unlink($file['tmp_name']);
+
+		if (!$content) {
+			throw new InvalidArgumentException('empty_file');
+		}
+		return $content;
+	}
+
+	public function validateFileFormat(string $fileName, $content): void {
+		$mimeType = $this->mimeTypeDetector->detectString($content);
+		$allowed = self::DEFAULT_ALLOWED_RECORDING_FORMATS;
+		if (!array_key_exists($mimeType, $allowed)) {
+			throw new InvalidArgumentException('file_mimetype');
+		}
+
+		$extension = strtolower(pathinfo($fileName, PATHINFO_EXTENSION));
+		if (!$extension || !in_array($extension, $allowed[$mimeType])) {
+			throw new InvalidArgumentException('file_extension');
+		}
+	}
+
+	public function validateFileName(string $fileName): string {
+		$recordFileName = escapeshellcmd($fileName);
+		$recordFileName = pathinfo($recordFileName, PATHINFO_BASENAME);
+		if ($recordFileName !== $fileName) {
+			throw new InvalidArgumentException('file_name');
+		}
+		return $recordFileName;
```

Comment:
This is not needed and all done by:
https://github.com/nextcloud/server/blob/a78595a45f5df29531205c5985ac6300c3094c6f/lib/private/Files/Node/Folder.php#L188

I would accept a `basename($file['name'])` on the original place, but this method should be removed